src/python/gudhi/perslay.py

```diff
@@ -0,0 +1,263 @@
+    bc_inp = tf.expand_dims(inp, -1)
+    return 1/tf.pow(1+tf.reduce_sum(tf.multiply(tf.abs(bc_inp - mu), tf.abs(sg)), axis=2), al)
+
+def exponential_layer(inp, num_elements, mean_init, variance_init, mean_const, variance_const, train_vect):
+    """ Exponential PersLay """
+    dimension_before, num_pts = inp.shape[2].value, inp.shape[1].value
+    mu = tf.get_variable("m", shape=[1, 1, dimension_before, num_elements], initializer=mean_init, trainable=train_vect)      if not mean_const      else tf.get_variable("m", initializer=tf.reshape(mean_init, [1, 1, dimension_before, num_elements]))
+    sg = tf.get_variable("s", shape=[1, 1, dimension_before, num_elements], initializer=variance_init, trainable=train_vect)  if not variance_const  else tf.get_variable("s", initializer=tf.reshape(variance_init, [1, 1, dimension_before, num_elements]))
+    bc_inp = tf.expand_dims(inp, -1)
+    return tf.exp(tf.reduce_sum(-tf.multiply(tf.square(bc_inp - mu), tf.square(sg)), axis=2))
+
+# Vectorizations implementing persistence landscapes
+def landscape_layer(inp, num_samples, sample_init, sample_const, train_vect):
+    """ Landscape PersLay """
+    sp = tf.get_variable("s", shape=[1, 1, num_samples], initializer=sample_init, trainable=train_vect) if not sample_const else tf.get_variable("s", initializer=tf.reshape(sample_init, [1,1,num_samples]))
+    return np.sqrt(2) * tf.maximum( .5 * (inp[:, :, 1:2] - inp[:, :, 0:1]) - tf.abs(sp - .5 * (inp[:, :, 1:2] + inp[:, :, 0:1])), np.array([0]))
+
+# Vectorizations implementing Betti curves
+def betti_layer(inp, theta, num_samples, sample_init, sample_const, train_vect):
+    """ Betti PersLay """
+    sp = tf.get_variable("s", shape=[1, 1, num_samples], initializer=sample_init, trainable=train_vect) if not sample_const else tf.get_variable("s", initializer=tf.reshape(sample_init, [1,1,num_samples]))
+    X, Y = inp[:, :, 0:1], inp[:, :, 1:2]
+    return  1. / ( 1. + tf.exp( -theta * (.5*(Y-X) - tf.abs(sp - .5*(Y+X))) )  )
+
+# Vectorizations implementing persistence entropy
+def entropy_layer(inp, theta, num_samples, sample_init, sample_const, train_vect):
+    """ Entropy PersLay
+    WARNING: this function assumes that padding values are zero
+    """
+    bp_inp = tf.einsum("ijk,kl->ijl", inp, tf.constant(np.array([[1.,-1.],[0.,1.]], dtype=np.float32)))
+    sp = tf.get_variable("s", shape=[1, 1, num_samples], initializer=sample_init, trainable=train_vect) if not sample_const else tf.get_variable("s", initializer=tf.reshape(sample_init, [1,1,num_samples]))
+    L, X, Y = bp_inp[:, :, 1:2], bp_inp[:, :, 0:1], bp_inp[:, :, 0:1] + bp_inp[:, :, 1:2]
+    LN = tf.multiply(L, 1. / tf.expand_dims(tf.matmul(L[:,:,0], tf.ones([L.shape[1],1])), -1))
+    entropy_terms = tf.where(LN > 0., -tf.multiply(LN, tf.log(LN)), LN)
+    return  tf.multiply(entropy_terms, 1. / ( 1. + tf.exp( -theta * (.5*(Y-X) - tf.abs(sp - .5*(Y+X))) )  ))
```

Comment:
The implementation of several functions is hard to follow, could you add some comments along the way, so that future readers can understand what the magic incantations (with einsum for instance) yield?

Reply:
Agreed. Also, `einsum("ijk, kl-> ijl") is used multiple times, maybe it could be given a name? (e.g. `tfdot` ?)